File: src/AetherNet.Core/Uri/AetherUri.cs
// SPDX-License-Identifier: MIT

using System.Collections.Generic;
using System.Globalization;
using System.Text;
using AetherNet.Identity;

namespace AetherNet.Uri;

/// <summary>
/// Aether URI — the canonical addressing format for resources on the Aether mesh.
///
/// <h3>Grammar (ABNF, RFC 5234)</h3>
/// <code>
/// aether-uri   = "aether://" authority [ "/" path ] [ "?" query ] [ "#" fragment ]
/// authority    = aether-tag / uhid
/// aether-tag   = 5(crockford) [ "-" ] 5(crockford)         ; case-insensitive
/// uhid         = 64(HEXDIG)                                ; SHA-256 hex of public key
/// path         = path-segment *( "/" path-segment )
/// path-segment = 1*( unreserved / pct-encoded / sub-delims / ":" / "@" )
/// query        = query-param *( "&amp;" query-param )
/// query-param  = key [ "=" value ]
/// key          = 1*( unreserved / pct-encoded )
/// value        = *( unreserved / pct-encoded / sub-delims / ":" / "@" / "/" / "?" )
/// fragment     = *( unreserved / pct-encoded / sub-delims / ":" / "@" / "/" / "?" )
/// crockford    = %x30-39 / %x41-48 / %x4A / %x4B / %x4D / %x4E / %x50-54 / %x56-5A
///              ; 0–9 A-H J K M N P-T V-Z (no I L O U)
/// unreserved   = ALPHA / DIGIT / "-" / "." / "_" / "~"
/// pct-encoded  = "%" HEXDIG HEXDIG
/// sub-delims   = "!" / "$" / "&amp;" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
/// </code>
///
/// <h3>Components</h3>
/// <list type="bullet">
///   <item><description><b>Scheme</b> is always <c>aether</c>. Case-insensitive on parse, lowercase on emit.</description></item>
///   <item><description><b>Authority</b> identifies the destination — an <see cref="AetherNetTag"/>
///     (10 Crockford base-32 chars, dash optional) or a UHID (64 hex chars). Case-insensitive.</description></item>
///   <item><description><b>Path</b> is opaque to the protocol — it names a handler within the destination
///     (e.g. <c>/content/&lt;hash&gt;</c>, <c>/profile</c>, <c>/inbox</c>). Case-sensitive.</description></item>
///   <item><description><b>Query</b> carries handler arguments. Keys are case-insensitive, values are case-sensitive.</description></item>
///   <item><description><b>Fragment</b> is a client-side hint and is never transmitted over the wire
///     (e.g. <c>#t=1m30s</c> for a playback position).</description></item>
/// </list>
///
/// <h3>Examples</h3>
/// <code>
/// aether://KXJB7-MN2P4/profile
/// aether://KXJB7-MN2P4/content/sha256-abc123?codec=opus
/// aether://KXJB7MN2P4/stream/live?bitrate=hd#t=1m30s
/// aether://a1b2c3d4e5f6...64hex/inbox
/// </code>
///
/// <h3>Why no userinfo or port</h3>
/// The authority IS the user — there is no separate userinfo. Ports have no meaning in
/// mesh routing because the transport layer selects the carrier (BLE / Wi-Fi Direct /
/// NearLink / HTTP relay); a URI never picks one.
/// </summary>
public readonly struct AetherUri : IEquatable<AetherUri>
{
    /// <summary>The fixed scheme name — <c>aether</c>.</summary>
    public const string Scheme = "aether";

    private const string SchemePrefix = "aether://";

    // ── State ────────────────────────────────────────────────────────────────

    /// <summary>The destination authority (AetherTag or UHID), canonicalised to upper case.</summary>
    public string Authority { get; }

    /// <summary>The handler path, without the leading slash. Empty string means "root".</summary>
    public string Path { get; }

    /// <summary>Decoded query parameters. Keys are stored lower-case for case-insensitive lookup.</summary>
    public IReadOnlyDictionary<string, string> Query { get; }

    /// <summary>The fragment, with leading "#" stripped. Empty if none.</summary>
    public string Fragment { get; }

    /// <summary>Returns true if this is a properly-parsed value (not a default struct).</summary>
    public bool IsValid => !string.IsNullOrEmpty(Authority);

    private AetherUri(string authority, string path, IReadOnlyDictionary<string, string> query, string fragment)
    {
        Authority = authority;
        Path = path;
        Query = query;
        Fragment = fragment;
    }

    // ── Parsing ──────────────────────────────────────────────────────────────

    /// <summary>
    /// Parses an <c>aether://</c> URI. Throws <see cref="AetherUriException"/> on any
    /// syntactic violation. Use <see cref="TryParse"/> for non-throwing parses.
    /// </summary>
    public static AetherUri Parse(string input)
    {
        if (input is null) throw new AetherUriException("Input is null.");
        if (!TryParse(input, out var uri, out var error))
            throw new AetherUriException(error ?? "Invalid aether URI.");
        return uri;
    }

    /// <summary>
    /// Attempts to parse an <c>aether://</c> URI. Returns true on success and populates
    /// <paramref name="result"/>; returns false and sets <paramref name="error"/> on failure.
    /// </summary>
    public static bool TryParse(string input, out AetherUri result, out string? error)
    {
        result = default;
        error = null;

        if (string.IsNullOrEmpty(input))
        {
            error = "Input is null or empty.";
            return false;
        }

        // Scheme is case-insensitive per RFC 3986.
        if (input.Length < SchemePrefix.Length ||
            !input.AsSpan(0, SchemePrefix.Length).Equals(SchemePrefix.AsSpan(), StringComparison.OrdinalIgnoreCase))
        {
            error = $"Scheme must be '{Scheme}://'.";
            return false;
        }

        var rest = input.Substring(SchemePrefix.Length);

        // Split on fragment first (only one '#' is allowed).
        var fragmentSplit = rest.IndexOf('#');
        string fragment = string.Empty;
        if (fragmentSplit >= 0)
        {
            fragment = PercentDecode(rest.Substring(fragmentSplit + 1));
            rest = rest.Substring(0, fragmentSplit);
        }

        // Then query.
        var querySplit = rest.IndexOf('?');
        var query = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        if (querySplit >= 0)
        {
            var queryRaw = rest.Substring(querySplit + 1);
            rest = rest.Substring(0, querySplit);
            foreach (var pair in queryRaw.Split('&', StringSplitOptions.RemoveEmptyEntries))
            {
                var eq = pair.IndexOf('=');
                var key = eq >= 0 ? pair.Substring(0, eq) : pair;
                var value = eq >= 0 ? pair.Substring(eq + 1) : string.Empty;
                var decodedKey = PercentDecode(key);
                if (decodedKey.Length == 0)
                {
                    error = "Empty query parameter key.";
                    return false;
                }
                query[decodedKey] = PercentDecode(value);
            }
        }

        // Then path.
        var pathSplit = rest.IndexOf('/');
        string authorityRaw;
        string path;
        if (pathSplit >= 0)
        {
            authorityRaw = rest.Substring(0, pathSplit);
            path = rest.Substring(pathSplit + 1);
        }
        else
        {
            authorityRaw = rest;
            path = string.Empty;
        }

        if (authorityRaw.Length == 0)
        {
            error = "Authority is missing.";
            return false;
        }

        // Authority validation: either an AetherNetTag or a 64-char hex UHID.
        var authority = CanonicaliseAuthority(authorityRaw, out var authError);
        if (authority is null)
        {
            error = authError;
            return false;
        }

        // Path validation: segments must contain only allowed characters.
        if (!ValidatePath(path, out var pathError))
        {
            error = pathError;
            return false;
        }

        result = new AetherUri(authority, PercentDecodePath(path), query, fragment);
        return true;
    }

    private static string? CanonicaliseAuthority(string raw, out string? error)
    {
        error = null;

        // Try UHID (64 hex chars).
        if (raw.Length == 64 && IsHex(raw))
        {
            return raw.ToUpperInvariant();
        }

        // Try AetherTag (10 Crockford chars with optional dash).
        if (AetherNetTag.TryParse(raw, out var tag))
        {
            return tag.Value; // Canonical form: XXXXX-XXXXX
        }

        error = $"Authority '{raw}' is neither a valid AetherTag nor a 64-char hex UHID.";
        return null;
    }

    private static bool IsHex(string s)
    {
        foreach (var c in s)
        {
            if (!((c >= '0' && c <= '9') ||
                  (c >= 'A' && c <= 'F') ||
                  (c >= 'a' && c <= 'f')))
                return false;
        }
        return true;
    }

    private static bool ValidatePath(string path, out string? error)
    {
        error = null;
        if (path.Length == 0) return true;

        // Walk segments, allow unreserved + pct-encoded + sub-delims + ':' + '@'.
        foreach (var segment in path.Split('/'))
        {
            if (segment.Length == 0)
            {
                error = "Empty path segment (consecutive slashes).";
                return false;
            }
            for (var i = 0; i < segment.Length; i++)
            {
                var c = segment[i];
                if (IsUnreserved(c) || IsSubDelim(c) || c == ':' || c == '@') continue;
                if (c == '%')
                {
                    if (i + 2 >= segment.Length || !IsHexChar(segment[i + 1]) || !IsHexChar(segment[i + 2]))
                    {
                        error = $"Malformed percent-encoding at position {i} of segment '{segment}'.";
                        return false;
                    }
                    i += 2;
                    continue;
                }
                error = $"Illegal character '{c}' in path segment '{segment}'.";
                return false;
            }
        }
        return true;
    }

    private static bool IsUnreserved(char c) =>
        (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') ||
        c == '-' || c == '.' || c == '_' || c == '~';

    private static bool IsSubDelim(char c) =>
        c == '!' || c == '$' || c == '&' || c == '\'' || c == '(' || c == ')' ||
        c == '*' || c == '+' || c == ',' || c == ';' || c == '=';

    private static bool IsHexChar(char c) =>
        (c >= '0' && c <= '9') || (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f');

    // ── Path helpers ─────────────────────────────────────────────────────────

    /// <summary>
    /// Returns the path split into segments (after percent-decoding each segment).
    /// Returns an empty array for the root path.
    /// </summary>
    public IReadOnlyList<string> PathSegments
    {
        get
        {
            if (string.IsNullOrEmpty(Path)) return Array.Empty<string>();
            var raw = Path.Split('/');
            var result = new string[raw.Length];
            for (var i = 0; i < raw.Length; i++)
                result[i] = raw[i]; // Path is already decoded.
            return result;
        }
    }

    /// <summary>
    /// Returns the first path segment (the "handler name"), or empty string for root.
    /// </summary>
    public string HandlerName
    {
        get
        {
            if (string.IsNullOrEmpty(Path)) return string.Empty;
            var slash = Path.IndexOf('/');
            return slash >= 0 ? Path.Substring(0, slash) : Path;
        }
    }

    // ── Encoding ─────────────────────────────────────────────────────────────

    /// <summary>
    /// Returns the canonical string form of this URI.
    /// Same value for two URIs that compare equal.
    /// </summary>
    public override string ToString()
    {
        if (!IsValid) return string.Empty;

        var sb = new StringBuilder(64);
        sb.Append(SchemePrefix);
        sb.Append(Authority);
        if (!string.IsNullOrEmpty(Path))
        {
            sb.Append('/');
            // Re-encode the path so the output is RFC-safe.
            EncodePath(sb, Path);
        }
        if (Query.Count > 0)
        {
            sb.Append('?');
            var first = true;
            foreach (var kv in Query)
            {
                if (!first) sb.Append('&');
                first = false;
                EncodeComponent(sb, kv.Key, EncodeKind.QueryKey);
                if (!string.IsNullOrEmpty(kv.Value))
                {
                    sb.Append('=');
                    EncodeComponent(sb, kv.Value, EncodeKind.QueryValue);
                }
            }
        }
        if (!string.IsNullOrEmpty(Fragment))
        {
            sb.Append('#');
            EncodeComponent(sb, Fragment, EncodeKind.Fragment);
        }
        return sb.ToString();
    }

    private enum EncodeKind { PathSegment, QueryKey, QueryValue, Fragment }

    private static void EncodePath(StringBuilder sb, string path)
    {
        var first = true;
        foreach (var segment in path.Split('/'))
        {
            if (!first) sb.Append('/');
            first = false;
            EncodeComponent(sb, segment, EncodeKind.PathSegment);
        }
    }

    private static void EncodeComponent(StringBuilder sb, string value, EncodeKind kind)
    {
        foreach (var c in value)
        {
            if (IsAllowedUnencoded(c, kind)) { sb.Append(c); continue; }
            // Percent-encode UTF-8 bytes.
            var bytes = Encoding.UTF8.GetBytes(new[] { c });
            foreach (var b in bytes)
            {
                sb.Append('%');
                sb.Append(b.ToString("X2", CultureInfo.InvariantCulture));
            }
        }
    }

    private static bool IsAllowedUnencoded(char c, EncodeKind kind)
    {
        if (IsUnreserved(c)) return true;
        switch (kind)
        {
            case EncodeKind.PathSegment:
                // pchar = unreserved / pct-encoded / sub-delims / ":" / "@"
                return IsSubDelim(c) || c == ':' || c == '@';
            case EncodeKind.QueryKey:
                // We always encode '&' and '=' in keys; allow ':' and '@' and the
                // other sub-delims that don't collide with the query syntax.
                return c == ':' || c == '@' ||
                       c == '!' || c == '$' || c == '\'' || c == '(' || c == ')' ||
                       c == '*' || c == '+' || c == ',' || c == ';';
            case EncodeKind.QueryValue:
                // Allow sub-delims except '&' (separator); '=' is fine inside a value.
                return c == ':' || c == '@' || c == '/' || c == '?' ||
                       c == '!' || c == '$' || c == '\'' || c == '(' || c == ')' ||
                       c == '*' || c == '+' || c == ',' || c == ';' || c == '=';
            case EncodeKind.Fragment:
                // fragment = *( pchar / "/" / "?" )  ; pchar incl. ':' '@' sub-delims
                return IsSubDelim(c) || c == ':' || c == '@' || c == '/' || c == '?';
            default:
                return false;
        }
    }

    private static string PercentDecode(string input)
    {
        if (input.IndexOf('%') < 0) return input;

        var bytes = new List<byte>(input.Length);
        for (var i = 0; i < input.Length; i++)
        {
            var c = input[i];
            if (c == '%' && i + 2 < input.Length &&
                IsHexChar(input[i + 1]) && IsHexChar(input[i + 2]))
            {
                bytes.Add((byte)((HexValue(input[i + 1]) << 4) | HexValue(input[i + 2])));
                i += 2;
            }
            else
            {
                // Non-encoded character — emit its UTF-8 bytes.
                foreach (var b in Encoding.UTF8.GetBytes(new[] { c }))
                    bytes.Add(b);
            }
        }
        return Encoding.UTF8.GetString(bytes.ToArray());
    }

    private static string PercentDecodePath(string path)
    {
        if (string.IsNullOrEmpty(path)) return path;
        // Decode each segment independently so '/' isn't lost.
        var segs = path.Split('/');
        for (var i = 0; i < segs.Length; i++) segs[i] = PercentDecode(segs[i]);
        return string.Join("/", segs);
    }

    private static int HexValue(char c) =>
        c <= '9' ? c - '0' :
        c <= 'F' ? c - 'A' + 10 :
                   c - 'a' + 10;

    // ── Equality ─────────────────────────────────────────────────────────────

    public bool Equals(AetherUri other)
    {
        if (Authority != other.Authority) return false;
        if (Path != other.Path) return false;
        if (Fragment != other.Fragment) return false;
        if (Query.Count != other.Query.Count) return false;
        foreach (var kv in Query)
            if (!other.Query.TryGetValue(kv.Key, out var v) || v != kv.Value) return false;
        return true;
    }

    public override bool Equals(object? obj) => obj is AetherUri other && Equals(other);

    public override int GetHashCode() =>
        HashCode.Combine(Authority, Path, Fragment, Query.Count);

    public static bool operator ==(AetherUri a, AetherUri b) => a.Equals(b);
    public static bool operator !=(AetherUri a, AetherUri b) => !a.Equals(b);
}

File: kotlin/src/main/kotlin/aethernet/uri/AetherUri.kt
// SPDX-License-Identifier: MIT

package aethernet.uri

import aethernet.identity.AetherNetTag
import java.util.Locale

/**
 * Aether URI — the canonical addressing format for resources on the Aether mesh.
 *
 * ## Grammar (ABNF, RFC 5234)
 * ```
 * aether-uri   = "aether://" authority [ "/" path ] [ "?" query ] [ "#" fragment ]
 * authority    = aether-tag / uhid
 * aether-tag   = 5(crockford) [ "-" ] 5(crockford)         ; case-insensitive
 * uhid         = 64(HEXDIG)                                ; SHA-256 hex of public key
 * path         = path-segment *( "/" path-segment )
 * path-segment = 1*( unreserved / pct-encoded / sub-delims / ":" / "@" )
 * query        = query-param *( "&" query-param )
 * query-param  = key [ "=" value ]
 * key          = 1*( unreserved / pct-encoded )
 * value        = *( unreserved / pct-encoded / sub-delims / ":" / "@" / "/" / "?" )
 * fragment     = *( unreserved / pct-encoded / sub-delims / ":" / "@" / "/" / "?" )
 * crockford    = %x30-39 / %x41-48 / %x4A / %x4B / %x4D / %x4E / %x50-54 / %x56-5A
 *              ; 0–9 A-H J K M N P-T V-Z (no I L O U)
 * unreserved   = ALPHA / DIGIT / "-" / "." / "_" / "~"
 * pct-encoded  = "%" HEXDIG HEXDIG
 * sub-delims   = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
 * ```
 *
 * ## Components
 *
 * - **Scheme** is always `aether`. Case-insensitive on parse, lower-case on emit.
 * - **Authority** identifies the destination — an [AetherNetTag] (10 Crockford
 *   base-32 chars, dash optional) or a UHID (64 hex chars). Case-insensitive.
 * - **Path** is opaque to the protocol — it names a handler within the destination
 *   (e.g. `/content/<hash>`, `/profile`, `/inbox`). Case-sensitive.
 * - **Query** carries handler arguments. Keys are case-insensitive (stored lower-case
 *   for lookup), values are case-sensitive.
 * - **Fragment** is a client-side hint and is never transmitted over the wire.
 *
 * ## Equality
 *
 * Two URIs are equal when their authority, path, fragment, and query map are all
 * equal. Query map equality is order-insensitive — `?a=1&b=2` equals `?b=2&a=1`.
 *
 * ## Examples
 * ```
 * aether://KXJB7-MN2P4/profile
 * aether://KXJB7-MN2P4/content/sha256-abc123?codec=opus
 * aether://KXJB7MN2P4/stream/live?bitrate=hd#t=1m30s
 * aether://a1b2c3d4e5f6...64hex/inbox
 * ```
 *
 * @property authority Destination authority (AetherTag or UHID), canonicalised to upper case.
 * @property path Handler path without the leading slash. Empty string means "root".
 * @property query Decoded query parameters. Keys are stored lower-case for case-insensitive lookup.
 * @property fragment The fragment, with leading "#" stripped. Empty if none.
 */
class AetherUri internal constructor(
    val authority: String,
    val path: String,
    val query: Map<String, String>,
    val fragment: String,
) {

    /** First path segment (handler name). Empty string for root paths. */
    val handlerName: String
        get() {
            if (path.isEmpty()) return ""
            val slash = path.indexOf('/')
            return if (slash >= 0) path.substring(0, slash) else path
        }

    /**
     * Path split into segments (already percent-decoded). Empty list for the root path.
     */
    val pathSegments: List<String>
        get() = if (path.isEmpty()) emptyList() else path.split('/')

    // ── toString — canonical encoder ─────────────────────────────────────────

    /**
     * Returns the canonical string form of this URI. Two URIs that compare equal
     * produce the same canonical string (modulo query-order, which is preserved
     * from the original parse / builder insertion order).
     */
    override fun toString(): String {
        val sb = StringBuilder(64)
        sb.append(SCHEME_PREFIX)
        sb.append(authority)
        if (path.isNotEmpty()) {
            sb.append('/')
            encodePath(sb, path)
        }
        if (query.isNotEmpty()) {
            sb.append('?')
            var first = true
            for ((k, v) in query) {
                if (!first) sb.append('&')
                first = false
                encodeComponent(sb, k, EncodeKind.QUERY_KEY)
                if (v.isNotEmpty()) {
                    sb.append('=')
                    encodeComponent(sb, v, EncodeKind.QUERY_VALUE)
                }
            }
        }
        if (fragment.isNotEmpty()) {
            sb.append('#')
            encodeComponent(sb, fragment, EncodeKind.FRAGMENT)
        }
        return sb.toString()
    }

    // ── Equality (query order-insensitive) ───────────────────────────────────

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is AetherUri) return false
        if (authority != other.authority) return false
        if (path != other.path) return false
        if (fragment != other.fragment) return false
        if (query.size != other.query.size) return false
        for ((k, v) in query) {
            if (other.query[k] != v) return false
        }
        return true
    }

    override fun hashCode(): Int {
        // Combine the same fields the C# struct does. Query is summarised by its
        // size so insertion order doesn't affect the hash — equal URIs always hash
        // to the same value.
        var h = authority.hashCode()
        h = 31 * h + path.hashCode()
        h = 31 * h + fragment.hashCode()
        h = 31 * h + query.size
        return h
    }

    /**
     * Result of a non-throwing [AetherUri.tryParse] call.
     *
     * Use the sealed class form for pattern matching:
     * ```
     * when (val r = AetherUri.tryParse(s)) {
     *     is AetherUri.ParseResult.Ok  -> use(r.uri)
     *     is AetherUri.ParseResult.Err -> log(r.message)
     * }
     * ```
     */
    sealed class ParseResult {
        data class Ok(val uri: AetherUri) : ParseResult()
        data class Err(val message: String) : ParseResult()
    }

    companion object {
        /** The fixed scheme name — `aether`. */
        const val SCHEME = "aether"

        private const val SCHEME_PREFIX = "aether://"

        // ── Parsing ──────────────────────────────────────────────────────────

        /**
         * Parses an `aether://` URI.
         *
         * @throws AetherUriException on any syntactic violation.
         */
        fun parse(input: String): AetherUri =
            when (val r = tryParse(input)) {
                is ParseResult.Ok -> r.uri
                is ParseResult.Err -> throw AetherUriException(r.message)
            }

        /**
         * Attempts to parse an `aether://` URI. Returns a sealed [ParseResult]
         * carrying either the parsed URI or the failure message.
         */
        fun tryParse(input: String): ParseResult {
            if (input.isEmpty()) {
                return ParseResult.Err("Input is null or empty.")
            }

            // Scheme is case-insensitive per RFC 3986.
            if (input.length < SCHEME_PREFIX.length ||
                !input.substring(0, SCHEME_PREFIX.length)
                    .equals(SCHEME_PREFIX, ignoreCase = true)
            ) {
                return ParseResult.Err("Scheme must be '$SCHEME://'.")
            }

            var rest = input.substring(SCHEME_PREFIX.length)

            // Split on fragment first (only one '#' is allowed).
            var fragment = ""
            val fragmentSplit = rest.indexOf('#')
            if (fragmentSplit >= 0) {
                fragment = percentDecode(rest.substring(fragmentSplit + 1))
                rest = rest.substring(0, fragmentSplit)
            }

            // Then query. We use a LinkedHashMap to preserve insertion order so
            // toString() output is deterministic per input.
            val query = LinkedHashMap<String, String>()
            val querySplit = rest.indexOf('?')
            if (querySplit >= 0) {
                val queryRaw = rest.substring(querySplit + 1)
                rest = rest.substring(0, querySplit)
                for (pair in queryRaw.split('&')) {
                    if (pair.isEmpty()) continue
                    val eq = pair.indexOf('=')
                    val rawKey = if (eq >= 0) pair.substring(0, eq) else pair
                    val rawValue = if (eq >= 0) pair.substring(eq + 1) else ""
                    val decodedKey = percentDecode(rawKey)
                    if (decodedKey.isEmpty()) {
                        return ParseResult.Err("Empty query parameter key.")
                    }
                    // Keys are case-insensitive — store lower-case so lookup works
                    // regardless of how the caller spelled them.
                    query[decodedKey.lowercase(Locale.ROOT)] = percentDecode(rawValue)
                }
            }

            // Then path.
            val pathSplit = rest.indexOf('/')
            val authorityRaw: String
            val path: String
            if (pathSplit >= 0) {
                authorityRaw = rest.substring(0, pathSplit)
                path = rest.substring(pathSplit + 1)
            } else {
                authorityRaw = rest
                path = ""
            }

            if (authorityRaw.isEmpty()) {
                return ParseResult.Err("Authority is missing.")
            }

            // Authority validation: either an AetherNetTag or a 64-char hex UHID.
            val authority = canonicaliseAuthority(authorityRaw)
                ?: return ParseResult.Err(
                    "Authority '$authorityRaw' is neither a valid AetherTag nor a 64-char hex UHID."
                )

            // Path validation: segments must contain only allowed characters.
            val pathErr = validatePath(path)
            if (pathErr != null) return ParseResult.Err(pathErr)

            return ParseResult.Ok(AetherUri(authority, percentDecodePath(path), query, fragment))
        }

        private fun canonicaliseAuthority(raw: String): String? {
            // Try UHID (64 hex chars).
            if (raw.length == 64 && isHex(raw)) {
                return raw.uppercase(Locale.ROOT)
            }
            // Try AetherTag (10 Crockford chars with optional dash).
            return AetherNetTag.tryParse(raw)?.value
        }

        private fun isHex(s: String): Boolean {
            for (c in s) {
                if (!((c in '0'..'9') || (c in 'A'..'F') || (c in 'a'..'f'))) return false
            }
            return true
        }

        private fun validatePath(path: String): String? {
            if (path.isEmpty()) return null

            // Walk segments, allow unreserved + pct-encoded + sub-delims + ':' + '@'.
            for (segment in path.split('/')) {
                if (segment.isEmpty()) {
                    return "Empty path segment (consecutive slashes)."
                }
                var i = 0
                while (i < segment.length) {
                    val c = segment[i]
                    if (isUnreserved(c) || isSubDelim(c) || c == ':' || c == '@') {
                        i++
                        continue
                    }
                    if (c == '%') {
                        if (i + 2 >= segment.length ||
                            !isHexChar(segment[i + 1]) ||
                            !isHexChar(segment[i + 2])
                        ) {
                            return "Malformed percent-encoding at position $i of segment '$segment'."
                        }
                        i += 3
                        continue
                    }
                    return "Illegal character '$c' in path segment '$segment'."
                }
            }
            return null
        }

        private fun isUnreserved(c: Char): Boolean =
            (c in 'A'..'Z') || (c in 'a'..'z') || (c in '0'..'9') ||
                c == '-' || c == '.' || c == '_' || c == '~'

        private fun isSubDelim(c: Char): Boolean =
            c == '!' || c == '$' || c == '&' || c == '\'' || c == '(' || c == ')' ||
                c == '*' || c == '+' || c == ',' || c == ';' || c == '='

        private fun isHexChar(c: Char): Boolean =
            (c in '0'..'9') || (c in 'A'..'F') || (c in 'a'..'f')

        // ── Percent codec ────────────────────────────────────────────────────

        private fun percentDecode(input: String): String {
            if (input.indexOf('%') < 0) return input

            val bytes = ArrayList<Byte>(input.length)
            var i = 0
            while (i < input.length) {
                val c = input[i]
                if (c == '%' && i + 2 < input.length &&
                    isHexChar(input[i + 1]) && isHexChar(input[i + 2])
                ) {
                    bytes.add(((hexValue(input[i + 1]) shl 4) or hexValue(input[i + 2])).toByte())
                    i += 3
                } else {
                    // Non-encoded character — emit its UTF-8 bytes.
                    for (b in c.toString().toByteArray(Charsets.UTF_8)) bytes.add(b)
                    i++
                }
            }
            return String(bytes.toByteArray(), Charsets.UTF_8)
        }

        private fun percentDecodePath(path: String): String {
            if (path.isEmpty()) return path
            // Decode each segment independently so '/' isn't lost.
            return path.split('/').joinToString("/") { percentDecode(it) }
        }

        private fun hexValue(c: Char): Int = when {
            c <= '9' -> c - '0'
            c <= 'F' -> c - 'A' + 10
            else -> c - 'a' + 10
        }
    }
}

/** Component-kind selector for the percent-encoder. */
private enum class EncodeKind { PATH_SEGMENT, QUERY_KEY, QUERY_VALUE, FRAGMENT }

private fun encodePath(sb: StringBuilder, path: String) {
    var first = true
    for (segment in path.split('/')) {
        if (!first) sb.append('/')
        first = false
        encodeComponent(sb, segment, EncodeKind.PATH_SEGMENT)
    }
}

private fun encodeComponent(sb: StringBuilder, value: String, kind: EncodeKind) {
    for (c in value) {
        if (isAllowedUnencoded(c, kind)) {
            sb.append(c)
            continue
        }
        // Percent-encode UTF-8 bytes.
        for (b in c.toString().toByteArray(Charsets.UTF_8)) {
            sb.append('%')
            sb.append("%02X".format(b.toInt() and 0xFF))
        }
    }
}

private fun isAllowedUnencoded(c: Char, kind: EncodeKind): Boolean {
    if (isUnreservedChar(c)) return true
    return when (kind) {
        // pchar = unreserved / pct-encoded / sub-delims / ":" / "@"
        EncodeKind.PATH_SEGMENT -> isSubDelimChar(c) || c == ':' || c == '@'
        // Always encode '&' and '=' in keys; allow ':' '@' and the other sub-delims
        // that don't collide with the query syntax.
        EncodeKind.QUERY_KEY -> c == ':' || c == '@' ||
            c == '!' || c == '$' || c == '\'' || c == '(' || c == ')' ||
            c == '*' || c == '+' || c == ',' || c == ';'
        // Allow sub-delims except '&' (separator); '=' is fine inside a value.
        EncodeKind.QUERY_VALUE -> c == ':' || c == '@' || c == '/' || c == '?' ||
            c == '!' || c == '$' || c == '\'' || c == '(' || c == ')' ||
            c == '*' || c == '+' || c == ',' || c == ';' || c == '='
        // fragment = *( pchar / "/" / "?" )  ; pchar incl. ':' '@' sub-delims
        EncodeKind.FRAGMENT -> isSubDelimChar(c) || c == ':' || c == '@' || c == '/' || c == '?'
    }
}

private fun isUnreservedChar(c: Char): Boolean =
    (c in 'A'..'Z') || (c in 'a'..'z') || (c in '0'..'9') ||
        c == '-' || c == '.' || c == '_' || c == '~'

private fun isSubDelimChar(c: Char): Boolean =
    c == '!' || c == '$' || c == '&' || c == '\'' || c == '(' || c == ')' ||
        c == '*' || c == '+' || c == ',' || c == ';' || c == '='

/** Thrown by the AetherUri parser, builder, and manifest when input is invalid. */
class AetherUriException(message: String) : RuntimeException(message)
